netkan/netkan/repos.py

```diff
@@ -0,0 +1,62 @@
+import logging
+from pathlib import Path
+from git import Repo
+
+from .metadata import Netkan, Ckan, CkanGroup
+
+
```

Comment:
From the comment earlier, we could have a MetaRepo class, that both these classes inherit from.

We can then use a setter/getter for cloning the repo. Sorta like this:
```python
class MetaRepo:

    def __init__(self, git_repo):
        self._git_repo = None
    
    @property
    def git_repo(self):
        if not self._git_repo:
            self._git_repo = init_repo(self._netkan_remote, '/tmp/NetKAN', self.deep_clone)
        return self._git_repo


class NetkanRepo(MetaRepo):

    """
    Encapsulates all assumptions we make about the structure of NetKAN
    """
    
    NETKAN_DIR = 'NetKAN'
    UNFROZEN_SUFFIX = 'netkan'
    FROZEN_SUFFIX = 'frozen'
    NETKAN_GLOB = f'**/*.{UNFROZEN_SUFFIX}'

    @property
    def nk_dir(self):
        return Path(self.git_repo.working_dir, self.NETKAN_DIR)


class CkanMetaRepo(MetaRepo):

    """
    Encapsulates all assumptions we make about the structure of CKAN-meta
    """
    
    CKANMETA_GLOB = '**/*.ckan'

    @property
    def ckm_dir(self):
        return Path(self.git_repo.working_dir)
```
Then we can abstract our git interactions, making it easier to potentially replace GitPython

Reply:
The wrinkle with having the `*kanRepo` classes call `init_repo` (which is what I did at first) is making that work with the Indexer tests:

https://github.com/KSP-CKAN/NetKAN-Infra/blob/4b5e1d65f66e0209025fef0e5ca1e7c087d9427d/netkan/tests/indexer.py#L40-L50

The repo is set up by the test setup itself using `git` operations, and the `*kanRepo` class needs to be able to receive that and work with it. Hence having its `__init__` take a `Repo` as a parameter is convenient and powerful.

Reply:
A shared parent class `__init__` won't set `self.nk_dir` or `self.ckm_dir`, which the classes need to operate properly. So they would each need `__init__`s anyway, which would call the parent `__init__` and then set those vars. Which would be the same number of lines of code, plus the extra lines to define the parent class. I don't think there's enough common logic to make this worth it.

Reply:
> The repo is set up by the test setup itself using git operations, and the *kanRepo class needs to be able to receive that and work with it. Hence having its __init__ take a Repo as a parameter is convenient and powerful.

We could still achieve this. Set repo as a kwarg, default to `None`. During `__init__` do a check and set it if is populated. The property will just return it if it's populated.

> A shared parent class __init__ won't set self.nk_dir or self.ckm_dir, which the classes need to operate properly. So they would each need __init__s anyway, which would call the parent __init__ and then set those vars. Which would be the same number of lines of code, plus the extra lines to define the parent class. I don't think there's enough common logic to make this worth it.

`self.nk_dir` + `self.ckm_dir` are converted to properties in my example. I don't see why they need to be initialised during class instantiation. The only reason you wouldn't want to do this is performance and you can do something similar to my repo example.

Things like branching, push, pull, commit etc are all common to both classes. A lot of that logic is buried in individual classes and could serve us better to centralise.